plugin-server/src/worker/ingestion/event-pipeline/enrichExceptionEventStep.ts

```diff
@@ -0,0 +1,57 @@
+import { Counter } from 'prom-client'
+
+import { PreIngestionEvent } from '../../../types'
+import { EventPipelineRunner } from './runner'
+
+const EXTERNAL_FINGERPRINT_COUNTER = new Counter({
+    name: 'enrich_exception_events_external_fingerprint',
+    help: 'Counter for exceptions that already have a fingerprint',
+})
+
+const COULD_NOT_PARSE_STACK_TRACE_COUNTER = new Counter({
+    name: 'enrich_exception_events_could_not_parse_stack_trace',
+    help: 'Counter for exceptions where the stack trace could not be parsed',
+})
+
+const EXCEPTIONS_ENRICHED_COUNTER = new Counter({
+    name: 'enrich_exception_events_enriched',
+    help: 'Counter for exceptions that have been enriched',
+})
+
+export function enrichExceptionEventStep(
+    _runner: EventPipelineRunner,
+    event: PreIngestionEvent
+): Promise<PreIngestionEvent> {
+    if (event.event !== '$exception') {
+        return Promise.resolve(event)
+    }
+
+    const exceptionStack = event.properties['$exception_stack_trace_raw']
+    const fingerPrint = event.properties['$exception_fingerprint']
+    const type = event.properties['$exception_type'] || 'no-type-provided'
+    const message = event.properties['$exception_message']
+    let firstFunction: string | null = null
+
+    if (fingerPrint) {
+        EXTERNAL_FINGERPRINT_COUNTER.inc()
+        return Promise.resolve(event)
+    }
+
+    try {
+        const parsedStack = JSON.parse(exceptionStack)
+        if (parsedStack.length > 0) {
+            firstFunction = parsedStack[0].function
+        }
+    } catch (e) {
+        COULD_NOT_PARSE_STACK_TRACE_COUNTER.inc()
+    }
+
+    if (message && firstFunction) {
+        event.properties['$exception_fingerprint'] = `${type}__${message}__${firstFunction}`
+    } else if (message) {
+        event.properties['$exception_fingerprint'] = `${type}__${message}`
+    }
```

Comment:
If you want to get rid of the conditional here:
```suggestion
        event.properties['$exception_fingerprint'] = [type, message, firstFunction].filter(Boolean).join("__")
```

Reply:
Message tends to contain a lot of potential wildcards. Possibly worth leaving it out until we handle them but also one could argue that over grouping is better than under grouping

Reply:
Yep, I'm 75% hoping to use this to generate a bunch of test cases from the data we do have

I think the next chunky thing is to move exception ingestion to its own url, so we can add a plugin server deployment just for those events, so that we can risk doing slower things

but i'm only just realising that as i type 🙈